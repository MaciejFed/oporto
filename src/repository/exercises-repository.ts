/* eslint-disable quotes */

import { db } from './db';
import { assertNonNull } from '../common/common';

export enum Person {
  Eu = 'Eu',
  Tu = 'Tu',
  ElaEleVocê = 'Ela/Ele/Você',
  Nós = 'Nós',
  ElesElasVosēs = 'Eles/Elas/Vocēs'
}

export type Verb = {
  english: string;
  infinitive: VerbInfinitive;
  presentSimple: { [key in Person]: string };
  pastPerfect?: { [key in Person]: string };
  imperfect?: { [key in Person]: string };
};

export type WordTypes = Noun | Verb | Other | Adjective;

type NounGender = 'masculine' | 'feminine' | 'none';
export type AdjectiveGender = 'masculine' | 'feminine';

type SentenceType = 'question' | 'statement';

export type AdjectiveNumber = 'singular' | 'plural';

type PortugueseNoun = {
  word: string;
  plural?: string;
  gender: NounGender;
};

export type Adjective = {
  masculine: {
    singular: string;
    plural: string;
  };
  feminine: {
    singular: string;
    plural: string;
  };
  placing?: 'before' | 'after';
  english: string;
};

export type Noun = {
  english: string;
  portuguese: PortugueseNoun;
};

export type Sentence = {
  english: string;
  portuguese: string;
  sentenceType?: SentenceType;
};

export type Phrase = {
  english: string;
  portuguese: string;
};

export type GenderWord = {
  feminine: string;
  masculine: string;
};

export type Other = {
  english: string;
  portuguese: string;
};

export type OtherWithGender = {
  english: string;
  portuguese: {
    base?: string;
    singular: GenderWord;
    plural: GenderWord;
  };
};

export type Schema = {
  verbs: Verb[];
  nouns: Noun[];
  others: Other[];
  othersWithGender: OtherWithGender[];
  adjectives: Adjective[];
  sentences: Sentence[];
  phrases: Phrase[];
  fitIn: FitIn[];
};

export type FitIn = {
  prefix: string;
  answer: string;
  suffix: string;
  explanation?: string;
};

const verbs = [
  'ser',
  'estar',
  'ir',
  'ter',
  'querer',
  'ver',
  'fazer',
  'saber',
  'ler',
  'dizer',
  'poder',
  'falar',
  'comer',
  'abrir',
  'achar',
  'andar',
  'beber',
  'entrar',
  'morar',
  'olhar',
  'chamar-se',
  'trabalhar',
  'cumprimentar',
  'apresentar-se',
  'encontrar',
  'perguntar',
  'jogar',
  'escrever',
  'compreender',
  'conhecer',
  'resolver',
  'descer',
  'aquecer',
  'viver',
  'correr',
  'aprender',
  'esquecer',
  'usar',
  'almoçar',
  'telefonar',
  'pagar',
  'tomar',
  'ficar',
  'levantar-se',
  'comprar',
  'ensinar',
  'fechar',
  'gostar',
  'começar',
  'lavar',
  'acabar',
  'apanhar',
  'vestir',
  'dividir',
  'despir',
  'corrigir',
  'decidir',
  'conseguir',
  'sentir',
  'traduzir',
  'partir',
  'preferir',
  'discutir',
  'conduzir',
  'permitir',
  'organizar',
  'limpar',
  'pôr',
  'cuidar',
  'mandar',
  'encher',
  'pegar',
  'tocar',
  'ajudar',
  'passear',
  'acreditar',
  'acontecer',
  'adorar',
  'alcançar',
  'alterar',
  'aplicar',
  'assegurar',
  'assinar',
  'aumentar',
  'cair',
  'carregar',
  'chegar',
  'concentrar',
  'construir',
  'continuar',
  'conversar',
  'criar',
  'dançar',
  'dar',
  'definir',
  'determinar',
  'dormir',
  'enviar',
  'esperar',
  'estudar',
  'explicar',
  'fumar',
  'gastar',
  'lembrar',
  'mencionar',
  'mudar',
  'negar',
  'parar',
  'passar',
  'pensar',
  'poupar',
  'preparar',
  'quebrar',
  'realizar',
  'refletir',
  'sair',
  'seguir',
  'servir',
  'sugerir',
  'tentar',
  'virar',
  'vender',
  'trazer',
  'transformar-se',
  'transportar',
  'sofrer',
  'segurar',
  'significar',
  'salvar',
  'sarar',
  'prover',
  'romper',
  'permanecer',
  'precisar',
  'ouvir',
  'pedir',
  'oferecer',
  'obter',
  'ocorrer',
  'mover',
  'mostrar',
  'brincar',
  'viajar',
  'atravessar',
  'promover',
  'casar',
  'levar',
  'agarrar',
  'responder',
  'sorrir',
  'suspirar',
  'nadar',
  'voar',
  'soltar',
  'separar',
  'empurrar',
  'voltar',
  'abanar',
  'gritar',
  'subir',
  'envenenar',
  'cantar',
  'rir',
  'assustar',
  'trincar',
  'sentar',
  'adormecer',
  'acordar',
  'reparar',
  'defender',
  'fugir',
  'convencer',
  'coçar',
  'bater',
  'piscar',
  'deixar',
  'avistar',
  'saltar',
  'mexer',
  'atirar',
  'acertar',
  'puxar',
  'haver'
] as const;

export type VerbInfinitive = (typeof verbs)[number];

export function readAll() {
  return db;
}

class WordDatabase {
  public constructor(private readonly database: Schema = db) {}

  public verb(verbInfinitive: VerbInfinitive): Verb {
    return assertNonNull(db.verbs.find((verb) => verb.infinitive === verbInfinitive));
  }
}

export const wordDatabase = new WordDatabase();
