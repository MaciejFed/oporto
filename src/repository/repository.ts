import { db, Schema } from './db';

const database: Schema = db;

export function readAll() {
    return database;
}


