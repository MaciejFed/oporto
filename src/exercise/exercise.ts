import { Comparable } from '../common/common';
import { RatioRange } from '../service/progress';
export type ExerciseType =
  | 'VerbExercise'
  | 'NounTranslation'
  | 'OtherTranslation'
  | 'AdjectiveTranslation'
  | 'VerbTranslation'
  | 'SentenceTranslation'
  | 'FitInGap';

export const translationTypes: ExerciseType[] = [
  'NounTranslation',
  'VerbTranslation',
  'OtherTranslation',
  'AdjectiveTranslation',
  'SentenceTranslation'
];

export interface Exercise extends Comparable {
  exerciseType: ExerciseType;
  getExerciseBodyPrefix(): string;
  getExerciseBodySuffix(): string;
  getExerciseDescription(): string;
  getExerciseTranslation(): string | undefined;
  getCorrectAnswer(): string;
  checkAnswerCorrect(answer: string): boolean;
  getRepeatAnswerPhrase(): string;
  getMaxWantedProgress(): RatioRange | '100+';
  getMinimumAnswers(): number;
}
