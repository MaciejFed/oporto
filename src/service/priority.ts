import { Exercise, ExerciseType } from '../exercise/exercise';
import { getAllResults, getAllResultsForExercise, getAllResultsForExerciseType } from '../repository/resultRepository';
import fs from 'fs';
import { Result } from './result';

export const VALUE_EXERCISE_DONE_WRONG = 30;
export const VALUE_EXERCISE_DONE_CORRECT = -10;
export const VALUE_EXERCISE_NEVER_DONE = 25;
export const VALUE_EXERCISE_TYPE_NEVER_DONE = 100;
export const VALUE_EXERCISE_DONE_TODAY = -15;

export type PriorityName =
  | 'EXERCISE_NEVER_DONE'
  | 'EXERCISE_TYPE_NEVER_DONE'
  | 'EXERCISE_WRONG'
  | 'EXERCISE_CORRECT'
  | 'EXERCISE_DONE_TODAY'
  | 'NO_PRIORITY';

export type Priority = {
  exercise: Exercise;
  priorityName: PriorityName;
  priorityValue: number;
};

type PriorityCompiler = (exercise: Exercise, results: Result[]) => Priority[];

export function sortExercises(exercises: Exercise[]): Exercise[] {
  const exercisesWithProrities = exercises
    .map((ex) => {
      const priorityCompilers: PriorityCompiler[] = [
        exerciseNeverDone,
        exerciseTypeNeverDone,
        exerciseWrong,
        exerciseCorrect,
        exerciseDoneToday
      ];
      const combinedProrites = priorityCompilers
        .flatMap((priorityCompiler) => priorityCompiler(ex, getAllResults()))
        .reduce(
          (pevious, current) => {
            pevious.priorityNames.push(current.priorityName);
            pevious.priorityValueTotal = pevious.priorityValueTotal + current.priorityValue;
            return pevious;
          },
          {
            priorityNames: [''],
            priorityValueTotal: 0,
            exercise: ex
          }
        );
      combinedProrites.priorityNames = combinedProrites.priorityNames.filter(
        (name) => name !== '' && name !== 'NO_PRIORITY'
      );
      return combinedProrites;
    })
    .sort((a, b) => b.priorityValueTotal - a.priorityValueTotal);

  fs.writeFileSync('prio.json', JSON.stringify(exercisesWithProrities, null, 2));

  return exercisesWithProrities.map((ewp) => ewp.exercise);
}

export function exerciseNeverDone(exercise: Exercise, results: Result[]): Priority[] {
  return getAllResultsForExercise(results, exercise).length === 0
    ? [{ exercise, priorityName: 'EXERCISE_NEVER_DONE', priorityValue: VALUE_EXERCISE_NEVER_DONE }]
    : noPriority(exercise);
}

export function exerciseTypeNeverDone(exercise: Exercise, results: Result[]): Priority[] {
  return getAllResultsForExerciseType(results, exercise.exercsiseType).length === 0
    ? [{ exercise, priorityName: 'EXERCISE_TYPE_NEVER_DONE', priorityValue: VALUE_EXERCISE_TYPE_NEVER_DONE }]
    : noPriority(exercise);
}

export function exerciseWrong(exercise: Exercise, results: Result[]): Priority[] {
  return [getAllResultsForExercise(results, exercise)
    .filter((result) => !result.isCorrect)
    .reduce((previous) => {
       previous.priorityValue += VALUE_EXERCISE_DONE_WRONG;
       return previous;
    }, { 
      exercise: exercise, priorityName: 'EXERCISE_WRONG', priorityValue: 0 
    })];
}

export function exerciseCorrect(exercise: Exercise, results: Result[]): Priority[] {
  return results
    .filter((result) => result.isCorrect)
    .map(() =>
      Object.assign({
        exercise: exercise,
        priorityName: 'EXERCISE_CORRECT',
        priorityValue: VALUE_EXERCISE_DONE_CORRECT
      })
    );
}

function exerciseDoneToday(exercise: Exercise, results: Result[]): Priority[] {
  const resultsToday = results.filter((result) => new Date(result.date).toDateString() === new Date().toDateString());

  return resultsToday.map(() =>
    Object.assign({
      exercise: exercise,
      priorityName: 'EXERCISE_DONE_TODAY',
      priorityValue: VALUE_EXERCISE_DONE_TODAY
    })
  );
}

export function noPriority(exercise: Exercise): Priority[] {
  return [
    {
      exercise,
      priorityName: 'NO_PRIORITY',
      priorityValue: 0
    }
  ];
}
