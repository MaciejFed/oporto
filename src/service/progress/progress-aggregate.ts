import { ExerciseProgress, getSingleExerciseProgress } from './progress';
import { Result } from '../result';
import { Exercise } from '../../exercise/exercise';
import { logger } from '../../common/logger';

enum ProgressType {
  DONE = 'DONE',
  IN_PROGRESS = 'IN_PROGRESS',
  NEVER_DONE = 'NEVER_DONE'
}

export const progressExerciseTypes = [
  'NounTranslation',
  'OtherTranslation',
  'AdjectiveTranslation',
  'VerbTranslation',
  'VerbExercise'
] as const;

type ExerciseTypeKey = (typeof progressExerciseTypes)[number];

type ProgressDetails = {
  count: number;
  baseWords: string[];
};

type ExerciseProgressAggregate = {
  [K in ExerciseTypeKey]: Record<ProgressType, ProgressDetails>;
};

type WordProgressAggregate = {
  [P in ProgressType]: ProgressDetails;
};

type ProgressAggregate = {
  exercises: ExerciseProgressAggregate;
  words: WordProgressAggregate;
};

const emptyProgressAggregate: ExerciseProgressAggregate = progressExerciseTypes.reduce(
  (prev, curr) => ({
    ...prev,
    ...{
      [curr]: {
        [ProgressType.DONE]: {
          count: 0,
          baseWords: []
        },
        [ProgressType.IN_PROGRESS]: {
          count: 0,
          baseWords: []
        },
        [ProgressType.NEVER_DONE]: {
          count: 0,
          baseWords: []
        }
      }
    }
  }),
  {}
) as ExerciseProgressAggregate;

const emptyWordProgressAggregate: WordProgressAggregate = {
  [ProgressType.DONE]: {
    count: 0,
    baseWords: []
  },
  [ProgressType.IN_PROGRESS]: {
    count: 0,
    baseWords: []
  },
  [ProgressType.NEVER_DONE]: {
    count: 0,
    baseWords: []
  }
};

export function getProgressAggregate(results: Result[], exercises: Exercise[]): ProgressAggregate {
  const exercisesFiltered = exercises.filter((exercise) =>
    progressExerciseTypes.includes(exercise.exerciseType as ExerciseTypeKey)
  );
  const exercisesProgress = exercisesFiltered
    .map((exercise) => getSingleExerciseProgress(results, exercise))
    .sort((a, b) => a.ratio - b.ratio);

  const addProgress = (
    typeToDetails: Record<ProgressType, ProgressDetails>,
    ep: ExerciseProgress
  ): Record<ProgressType, ProgressDetails> => {
    const key =
      // eslint-disable-next-line no-nested-ternary
      ep.ratioRange === 'Never Done'
        ? ProgressType.NEVER_DONE
        : ep.ratioRange === '80-100'
        ? ProgressType.DONE
        : ProgressType.IN_PROGRESS;
    const baseWords = [...new Set(typeToDetails[key].baseWords.concat(ep.exercise.getBaseWordAsString() || ''))];
    return {
      ...typeToDetails,
      [key]: {
        count: baseWords.length,
        baseWords
      }
    };
  };

  const progressAggregate = exercisesProgress.reduce((prev, curr) => {
    return {
      ...prev,
      [curr.exercise.exerciseType]: {
        ...prev[curr.exercise.exerciseType as ExerciseTypeKey],
        ...addProgress(prev[curr.exercise.exerciseType as ExerciseTypeKey], curr)
      }
    };
  }, emptyProgressAggregate);

  const allBaseWords = [...new Set(exercisesFiltered.map((exercise) => exercise.getBaseWordAsString() || ''))];
  const wordProgressAggregate: WordProgressAggregate = allBaseWords.reduce((prev, curr) => {
    const wordExerciseProgress = exercisesProgress.filter((ep) => ep.exercise.getBaseWordAsString() === curr);
    const neverDoneCount = wordExerciseProgress.filter((ep) => ep.ratioRange === 'Never Done').length;
    const inProgressCount = wordExerciseProgress.filter(
      (ep) => ep.ratioRange === '0-39' || ep.ratioRange === '40-79'
    ).length;
    const doneProgressCount = wordExerciseProgress.filter((ep) => ep.ratioRange === '80-100').length;

    if (doneProgressCount > 0 && inProgressCount === 0 && neverDoneCount === 0) {
      return {
        ...prev,
        [ProgressType.DONE]: {
          baseWords: prev.DONE.baseWords.concat(curr),
          count: prev.DONE.count + 1
        }
      };
    }
    if (inProgressCount > 0 || (neverDoneCount > 0 && doneProgressCount > 0)) {
      return {
        ...prev,
        [ProgressType.IN_PROGRESS]: {
          baseWords: prev.IN_PROGRESS.baseWords.concat(curr),
          count: prev.IN_PROGRESS.count + 1
        }
      };
    }
    if (neverDoneCount > 0 && doneProgressCount === 0 && inProgressCount === 0) {
      return {
        ...prev,
        [ProgressType.NEVER_DONE]: {
          baseWords: prev.NEVER_DONE.baseWords.concat(curr),
          count: prev.NEVER_DONE.count + 1
        }
      };
    }
    logger.error('Unexpected state');
    return prev;
  }, emptyWordProgressAggregate);

  return {
    words: wordProgressAggregate,
    exercises: progressAggregate
  };
}
