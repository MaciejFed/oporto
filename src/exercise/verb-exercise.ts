import { Comparable } from '../common/common';
import { Person, Verb } from '../repository/exercises-repository';
import { getCorrectVerbConjugation, getRandomPerson, getRandomVerb } from '../service/verb';
import { Exercise, ExerciseType } from './exercise';
import { RatioRange } from '../service/progress';

export type VerbTime = 'presentSimple' | 'pastPerfect';

const vertTimePrintVersion = (verbTime: VerbTime) => {
  if (verbTime === 'presentSimple') return 'Present Simple';
  return 'Past Perfect';
};

export class VerbExercise implements Exercise, Comparable {
  exerciseType: ExerciseType;
  verbTime: VerbTime;
  verb: Verb;
  person: Person;

  constructor() {
    this.exerciseType = 'VerbExercise';
    this.verb = getRandomVerb();
    this.person = getRandomPerson();
    this.verbTime = 'presentSimple';
  }

  static new(verb: Verb, person: Person, verbTime: VerbTime): VerbExercise {
    const verbExercise = new VerbExercise();
    verbExercise.verb = verb;
    verbExercise.person = person;
    verbExercise.verbTime = verbTime;

    return verbExercise;
  }

  getExerciseTranslation = () => undefined;

  getExerciseBodyPrefix = () => `[${vertTimePrintVersion(this.verbTime)}] ${this.person}: `;

  getExerciseBodySuffix = () => '';

  getExerciseDescription = () => `Infinitive: ${this.verb.english}`;

  getCorrectAnswer = () => getCorrectVerbConjugation(this.verb, this.person, this.verbTime);

  checkAnswerCorrect(answer: string): boolean {
    const correctConjugation = this.getCorrectAnswer();
    return correctConjugation.toLowerCase() === answer.toLowerCase();
  }

  getMaxWantedProgress(): RatioRange | '100+' {
    return '80-100';
  }

  getMinimumAnswers(): number {
    return 2;
  }

  getRepeatAnswerPhrase = () => `${this.person} ${this.getCorrectAnswer()}`;

  equal = (exercise: VerbExercise) =>
    exercise.exerciseType === 'VerbExercise' &&
    this.verb.infinitive === exercise.verb.infinitive &&
    this.verbTime === exercise.verbTime &&
    this.person === exercise.person;
}
