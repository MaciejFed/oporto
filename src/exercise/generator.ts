import { Exercise } from './exercise';
import { Person, readAll } from '../repository/exercises-repository';
import { VerbExercise } from './verb-exercise';
import { NounTranslationExercise } from './translation/noun-translation-exercise';
import { TranslationType } from './translation/translation-exercise';
import { VerbTranslationExercise } from './translation/verb-translation-exercise';
import { AdjectiveTranslationExercise } from './translation/adjective-translation-exercise';
import { SentenceTranslationExercise } from './translation/sentence-translation-exercise';
import { FitInGapExercise } from './fit-in-gap-exercise';
import { OtherTranslationExercise } from './translation/other-translation-exercise';
import { sortExercises } from '../priority/priority';
import { PhraseTranslationExercise } from './translation/phrase-translation-exercise';
import { exerciseFactory, getAllResults, getAllResultsAsync, parseResults } from '../repository/result-repository';
import { fetchExercisesForSession } from '../client/client';
import { Result } from '../service/result';
import { checkStandardConjugation } from '../service/verb/verb';
import { Language } from '../common/language';
import {
  GenderWord,
  GermanCase,
  GermanGender,
  GermanPerson,
  readAllDE
} from '../repository/german-exercises-repository';
import { GermanNounTranslationExercise } from './translation/de/german-noun-translation-exercise';
import { GermanVerbTranslationExercise } from './translation/de/german-verb-translation-exercise';
import { GermanVerbExercise } from './german-verb-exercise';
import { GermanOtherTranslationExercise } from './translation/de/german-other-translation-exercise';
import { GermanCaseExercise } from './german-case-exercise';
import { PolishPerson, readAllPL } from '../repository/polish-exercises-repository';
import { PolishVerbTranslationExercise } from './translation/pl/polish-verb-translation-exercise';
import { PolishOtherTranslationExercise } from './translation/pl/polish-other-translation-exercise';
import { PolishNounTranslationExercise } from './translation/pl/polish-noun-translation-exercise';
import { PolishVerbExercise } from './polish-verb-exercise';

type ExerciseGenerator = () => Exercise[];

export const VerbExerciseGenerator: ExerciseGenerator = () => {
  const verbsNonStandard = readAll().verbs.filter((verb) => !checkStandardConjugation(verb.infinitive, []).isStandard);
  const presentSimpleVerbs = verbsNonStandard.flatMap((verb) =>
    Object.keys(Person).flatMap((person) =>
      VerbExercise.new(verb, Person[person as keyof typeof Person], 'presentSimple')
    )
  );
  const pastPerfectVerbs = verbsNonStandard
    .filter((verb) => verb.pastPerfect)
    .flatMap((verb) =>
      Object.keys(Person).flatMap((person) =>
        VerbExercise.new(verb, Person[person as keyof typeof Person], 'pastPerfect')
      )
    );
  return pastPerfectVerbs.concat(presentSimpleVerbs);
};

export const GermanVerbExerciseGenerator: ExerciseGenerator = () => {
  return readAllDE().verbs.flatMap((verb) =>
    Object.keys(GermanPerson).flatMap((person) => [
      ...[GermanVerbExercise.new(verb, GermanPerson[person as keyof typeof GermanPerson], 'presentSimple')],
      ...(verb.pastPerfect
        ? [GermanVerbExercise.new(verb, GermanPerson[person as keyof typeof GermanPerson], 'pastPerfect')]
        : []),
      ...(verb.präteritum
        ? [GermanVerbExercise.new(verb, GermanPerson[person as keyof typeof GermanPerson], 'präteritum')]
        : [])
    ])
  );
};

export const PolishVerbExerciseGenerator: ExerciseGenerator = () => {
  return readAllPL().verbs.flatMap((verb) =>
    Object.keys(PolishPerson).flatMap((person) => [PolishVerbExercise.new(verb, person as any)])
  );
};

const translationTypes: TranslationType[] = ['toPortugueseFromHearing', 'toEnglish', 'toPortuguese'];

const NounTranslationGenerator: ExerciseGenerator = () => {
  return readAll().nouns.flatMap((noun) =>
    translationTypes.map((translationType) => NounTranslationExercise.new(noun, translationType))
  );
};

const GermanNounTranslationGenerator: ExerciseGenerator = () => {
  return readAllDE().nouns.flatMap((noun) =>
    translationTypes.flatMap((translationType) => [
      ...[GermanNounTranslationExercise.new(noun, translationType, 'singular')],
      ...(noun.german.plural ? [GermanNounTranslationExercise.new(noun, translationType, 'plural')] : [])
    ])
  );
};

const PolishNounTranslationGenerator: ExerciseGenerator = () => {
  return readAllPL().nouns.flatMap((noun) =>
    translationTypes.map((translationType) => PolishNounTranslationExercise.new(noun, translationType))
  );
};

const GermanOtherTranslationGenerator: ExerciseGenerator = () => {
  return readAllDE().others.flatMap((other) =>
    translationTypes.map((translationType) => GermanOtherTranslationExercise.new(other, translationType))
  );
};

const PolishOtherTranslationGenerator: ExerciseGenerator = () => {
  return readAllPL().others.flatMap((other) =>
    translationTypes.map((translationType) => PolishOtherTranslationExercise.new(other, translationType))
  );
};

const GermanCaseWordGenerator: ExerciseGenerator = () => {
  const nonGenderWords = readAllDE().case.filter((caseWord) => typeof caseWord.german.nominative === 'string');
  const genderWords = readAllDE().case.filter((caseWord) => typeof caseWord.german.nominative === 'object');

  const exercises = Object.values(GermanCase)
    .flatMap((gemanCase) => [
      nonGenderWords.filter((word) => word.german[gemanCase]).map((word) => GermanCaseExercise.new(word, gemanCase)),
      ['maskulinum', 'femininum', 'neutrum', 'plural'].flatMap((gender) =>
        genderWords.flatMap((word) => {
          // @ts-ignore
          if (!word.german[gemanCase] || !word.german[gemanCase][gender]) return [];
          // @ts-ignore
          return [GermanCaseExercise.new(word, gemanCase, gender)];
        })
      )
    ])
    .flatMap((e) => e);

  return exercises;
};

const GermanVerbTranslationGenerator: ExerciseGenerator = () => {
  return readAllDE().verbs.flatMap((verb) =>
    translationTypes.map((translationType) => GermanVerbTranslationExercise.new(verb, translationType))
  );
};

const PolishVerbTranslationGenerator: ExerciseGenerator = () => {
  return readAllPL().verbs.flatMap((verb) =>
    translationTypes.map((translationType) => PolishVerbTranslationExercise.new(verb, translationType))
  );
};

const VerbTranslationGenerator: ExerciseGenerator = () => {
  return readAll().verbs.flatMap((verb) =>
    translationTypes.map((translationType) => VerbTranslationExercise.new(verb, translationType))
  );
};

const SentenceTranslationGenerator: ExerciseGenerator = () => {
  return readAll().sentences.flatMap((sentence) => [
    SentenceTranslationExercise.new(sentence, 'toPortugueseFromHearing'),
    SentenceTranslationExercise.new(sentence, 'toPortuguese')
  ]);
};

const PhraseTranslationGenerator: ExerciseGenerator = () => {
  return readAll().phrases.flatMap((phrase) => [
    PhraseTranslationExercise.new(phrase, 'toPortugueseFromHearing'),
    PhraseTranslationExercise.new(phrase, 'toPortuguese')
  ]);
};

const OtherTranslationGenerator: ExerciseGenerator = () => {
  return readAll().others.flatMap((other) =>
    translationTypes.map((translationType) => OtherTranslationExercise.new(other, translationType))
  );
};

const AdjectiveTranslationGenerator: ExerciseGenerator = () => {
  return readAll().adjectives.flatMap((noun) =>
    translationTypes.flatMap((translationType) => {
      if (translationType === 'toPortuguese') {
        return [
          AdjectiveTranslationExercise.new(noun, translationType, 'masculine', 'singular'),
          AdjectiveTranslationExercise.new(noun, translationType, 'masculine', 'plural'),
          AdjectiveTranslationExercise.new(noun, translationType, 'feminine', 'singular'),
          AdjectiveTranslationExercise.new(noun, translationType, 'feminine', 'plural')
        ];
      }
      return [AdjectiveTranslationExercise.new(noun, translationType, 'masculine', 'singular')];
    })
  );
};

const FitInGapGenerator: ExerciseGenerator = () => {
  return readAll().fitIn.map((fit) => {
    const fitInGapExercise = new FitInGapExercise();
    fitInGapExercise.fitIn = fit;

    return fitInGapExercise;
  });
};

export function generateAllPossibleExercises(language: Language): Exercise[] {
  switch (language) {
    case Language.German:
      return [
        GermanVerbExerciseGenerator,
        GermanNounTranslationGenerator,
        GermanVerbTranslationGenerator,
        GermanOtherTranslationGenerator,
        GermanCaseWordGenerator
      ].flatMap((generator) => generator());

    case Language.Polish:
      return [
        PolishVerbExerciseGenerator,
        PolishNounTranslationGenerator,
        PolishVerbTranslationGenerator,
        PolishOtherTranslationGenerator
      ].flatMap((generator) => generator());
    case Language.Portuguese:
    default:
      return [
        VerbExerciseGenerator,
        NounTranslationGenerator,
        VerbTranslationGenerator,
        PhraseTranslationGenerator,
        OtherTranslationGenerator,
        AdjectiveTranslationGenerator,
        FitInGapGenerator
      ].flatMap((generator) => generator());
  }
}

export async function generateExercisesForSessionAsync(
  exerciseCount: number,
  sort: boolean,
  filter: (ex: Exercise) => boolean,
  language: Language,
  results?: Result[]
): Promise<Exercise[]> {
  const exercises = generateAllPossibleExercises(language).filter((exercise) => filter(exercise));
  const allResults = results ? parseResults(results) : await getAllResultsAsync(language);
  const exercisesFinal = sort ? sortExercises(exercises, allResults, language).exercises : exercises;

  return exercisesFinal.splice(0, Math.min(exerciseCount, exercisesFinal.length - 1)).reverse();
}

export function getExercisesForSession(language: Language): Exercise[] {
  const exerciseJSON: Exercise[] = fetchExercisesForSession(language);
  const exercies = exerciseJSON.map((ex) => {
    const exerciseType = ex.exerciseType;
    const createExercise = exerciseFactory[exerciseType];
    return createExercise(ex);
  });
  return exercies;
}

export function generateExercisesForSession(
  exerciseCount: number,
  sort: boolean,
  filter: (ex: Exercise) => boolean,
  language: Language
): Exercise[] {
  const exercises = generateAllPossibleExercises(language).filter((exercise) => filter(exercise));
  const allResults = getAllResults(language);
  const exercisesFinal = sort ? sortExercises(exercises, allResults, language).exercises : exercises;

  return exercisesFinal.splice(0, Math.min(exerciseCount, exercisesFinal.length - 1)).reverse();
}
