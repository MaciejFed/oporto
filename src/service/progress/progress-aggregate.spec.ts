import { generateResultForExercise } from '../../priority/priority.util';
import { Person, wordDatabase } from '../../repository/exercises-repository';
import { VerbExercise } from '../../exercise/verb-exercise';
import { getProgressAggregate } from './progress-aggregate';
import {generateAllPossibleExercises} from "../../exercise/generator";

const conhecerVerbExerciseDone1 = VerbExercise.new(wordDatabase.verb('conhecer'), Person.Tu, 'presentSimple');
const saberVerbExerciseDone1 = VerbExercise.new(wordDatabase.verb('saber'), Person.Eu, 'pastPerfect');
const saberVerbExerciseInProgress1 = VerbExercise.new(wordDatabase.verb('saber'), Person.Tu, 'pastPerfect');
const verVerbExerciseNeverDone = VerbExercise.new(wordDatabase.verb('ver'), Person.Tu, 'pastPerfect');

const results = [
  ...generateResultForExercise(conhecerVerbExerciseDone1, true, 'keyboard', 1),
  ...generateResultForExercise(saberVerbExerciseDone1, true, 'keyboard', 1),
  ...generateResultForExercise(saberVerbExerciseInProgress1, false, 'keyboard', 1)
];

describe('Progress Aggregate', () => {
  it('getProgressAggregate', () => {
    const progressAggregate = getProgressAggregate(results, [
      conhecerVerbExerciseDone1,
      saberVerbExerciseDone1,
      saberVerbExerciseInProgress1,
      verVerbExerciseNeverDone
    ]);

    expect(progressAggregate).toMatchSnapshot();
  });

  it('getProgressAggregate full exercises', () => {
    const progressAggregate = getProgressAggregate(results, generateAllPossibleExercises());

    expect(progressAggregate).toMatchSnapshot();
  });
});
