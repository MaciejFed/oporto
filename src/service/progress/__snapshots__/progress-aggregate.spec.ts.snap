// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Aggregate getProgressAggregate 1`] = `
Object {
  "exercises": Object {
    "AdjectiveTranslation": Object {
      "DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "IN_PROGRESS": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "NEVER_DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
    },
    "NounTranslation": Object {
      "DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "IN_PROGRESS": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "NEVER_DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
    },
    "OtherTranslation": Object {
      "DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "IN_PROGRESS": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "NEVER_DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
    },
    "VerbExercise": Object {
      "DONE": Object {
        "baseWords": Array [
          "conhecer",
          "saber",
        ],
        "count": 2,
      },
      "IN_PROGRESS": Object {
        "baseWords": Array [
          "saber",
        ],
        "count": 1,
      },
      "NEVER_DONE": Object {
        "baseWords": Array [
          "ver",
        ],
        "count": 1,
      },
    },
    "VerbTranslation": Object {
      "DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "IN_PROGRESS": Object {
        "baseWords": Array [],
        "count": 0,
      },
      "NEVER_DONE": Object {
        "baseWords": Array [],
        "count": 0,
      },
    },
  },
  "words": Object {
    "DONE": Object {
      "baseWords": Array [
        "conhecer",
      ],
      "count": 1,
    },
    "IN_PROGRESS": Object {
      "baseWords": Array [
        "saber",
      ],
      "count": 1,
    },
    "NEVER_DONE": Object {
      "baseWords": Array [
        "ver",
      ],
      "count": 1,
    },
  },
}
`;
